File: algorithms/easy/35_search_insert_position.py
# 35. Search Insert Position

def search_insert(nums, target):
    """Find the index where target is found, or where it would be inserted in order.
    
    Args:
        nums: Sorted array of distinct integers
        target: Target value to find or insert
    
    Returns:
        Index where target is found, or where it would be inserted in order
    """
    left, right = 0, len(nums) - 1
    
    while left <= right:
        mid = left + (right - left) // 2
        
        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    
    # If target is not found, 'left' will be the position where it should be inserted
    return left

# Test cases
print(search_insert([1, 3, 5, 6], 5))      # 2
print(search_insert([1, 3, 5, 6], 2))      # 1
print(search_insert([1, 3, 5, 6], 7))      # 4
print(search_insert([1, 3, 5, 6], 0))      # 0

File: algorithms/easy/35_search_insert_position.js
/**
 * 35. Search Insert Position
 * 
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function(nums, target) {
    let left = 0;
    let right = nums.length - 1;
    
    while (left <= right) {
        const mid = left + Math.floor((right - left) / 2);
        
        if (nums[mid] === target) {
            return mid;
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    
    // If target is not found, 'left' will be the position where it should be inserted
    return left;
};

// Test cases
console.log(searchInsert([1, 3, 5, 6], 5));      // 2
console.log(searchInsert([1, 3, 5, 6], 2));      // 1
console.log(searchInsert([1, 3, 5, 6], 7));      // 4
console.log(searchInsert([1, 3, 5, 6], 0));      // 0